test/distributed/org/apache/cassandra/distributed/test/sai/PartialUpdateHandlingTest.java

```diff
@@ -0,0 +1,515 @@
+ *     fail to return live matches with partial updates and no previous data.
+ * 3.) The number of query clauses and their targets. Clauses may target partition keys, clustering keys, static 
+ *     columns, and regular columns, and some combinations are more problematic than others.
+ * 4.) The repaired state of SSTables that participate in the query. A fully repaired on-disk set of SSTables cannot
+ *     produce erroneous results due to split rows (i.e. rows which contain partial updates of different columns
+ *     across different partitions).
+ * 5.) Whether data resides in SSTables or Memtables. The latter is implicitly unrepaired.
+ * 6.) Interaction w/ existing mechanisms on the distributed read path that deal with short reads, replica filtering
+ *     protection, etc.
+ * 7.) The relationship between columns selected and columns restricted by queries. (If coordinator filtering is
+ *     involved at the implementation level, retrieving enough information to do that filtering is important.)
+ * 8.) The timestamps of partial updates and deletes, especially for single-column queries that might produce
+ *     stale matches if not resolved correctly.
+ */
+@RunWith(Parameterized.class)
+public class PartialUpdateHandlingTest extends TestBaseImpl
+{
+    private static final String TEST_TABLE_NAME = "test_partial_updates";
+    private static final int PARTITIONS_PER_TEST = 10;
+    private static final int NODES = 3;
+
+    private static Cluster CLUSTER;
+
+    @BeforeClass
+    public static void setUpCluster() throws IOException
+    {
+        CLUSTER = init(Cluster.build(NODES).withConfig(config -> config.set("hinted_handoff_enabled", false).with(GOSSIP).with(NETWORK)).start());
+
+        // All parameterized test scenarios share the same table and attached indexes, but write to different partitions
+        // that are deleted after each scenario completes.
+        String createTableDDL = String.format("CREATE TABLE %s.%s (pk int, pk2 int, ck int, s int static, a int, b int, PRIMARY KEY ((pk, pk2), ck)) WITH read_repair = 'NONE'",
+                                              KEYSPACE, TEST_TABLE_NAME);
+        CLUSTER.schemaChange(createTableDDL);
+        CLUSTER.disableAutoCompaction(KEYSPACE);
+
+        CLUSTER.schemaChange(String.format("CREATE INDEX pk2_idx ON %s.%s(pk2) USING 'sai'", KEYSPACE, TEST_TABLE_NAME));
+        CLUSTER.schemaChange(String.format("CREATE INDEX ck_idx ON %s.%s(ck) USING 'sai'", KEYSPACE, TEST_TABLE_NAME));
+        CLUSTER.schemaChange(String.format("CREATE INDEX s_idx ON %s.%s(s) USING 'sai'", KEYSPACE, TEST_TABLE_NAME));
+        CLUSTER.schemaChange(String.format("CREATE INDEX a_idx ON %s.%s(a) USING 'sai'", KEYSPACE, TEST_TABLE_NAME));
+        CLUSTER.schemaChange(String.format("CREATE INDEX b_idx ON %s.%s(b) USING 'sai'", KEYSPACE, TEST_TABLE_NAME));
+
+        SAIUtil.waitForIndexQueryable(CLUSTER, KEYSPACE);
```

Comment:
Locally, I've actually also tested basic filtering w/ this test matrix. Really all you have to do it eliminate these index creations, comment out the index queryability stuff, and add ALLOW FILTERING to the model validation SELECT, and you fail about 44 of the 184 tests (although it would be more if we forced node 2 to be ignored on QUORUM validations)

Reply:
See https://github.com/maedhroz/cassandra/pull/14